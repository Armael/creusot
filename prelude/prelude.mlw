module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
