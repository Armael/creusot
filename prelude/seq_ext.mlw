module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end