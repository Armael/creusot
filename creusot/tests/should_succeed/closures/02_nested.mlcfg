
module C02Nested_NestedClosure_Closure0_Closure0_Type
  use prelude.Int8
  use prelude.Borrow
  type c02nested_nestedclosure_closure0_closure0  =
    | C02Nested_NestedClosure_Closure0_Closure0 bool
    
end
module C02Nested_NestedClosure_Closure0_Closure0
  use prelude.Int8
  use C02Nested_NestedClosure_Closure0_Closure0_Type as C02Nested_NestedClosure_Closure0_Closure0
  use prelude.Borrow
  let function field_00 [#"../02_nested.rs" 6 18 6 20] (self : C02Nested_NestedClosure_Closure0_Closure0.c02nested_nestedclosure_closure0_closure0) : bool
    
   = [@vc:do_not_keep_trace] [@vc:sp]
    [#"../02_nested.rs" 1 0 1 0] let C02Nested_NestedClosure_Closure0_Closure0.C02Nested_NestedClosure_Closure0_Closure0 a = self in a
  let rec cfg c02Nested_NestedClosure_Closure0_Closure0 [#"../02_nested.rs" 6 18 6 20] [@cfg:stackify] [@cfg:subregion_analysis] (_1 : C02Nested_NestedClosure_Closure0_Closure0.c02nested_nestedclosure_closure0_closure0) : bool
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : bool;
  var _1 : C02Nested_NestedClosure_Closure0_Closure0.c02nested_nestedclosure_closure0_closure0 = _1;
  {
    goto BB0
  }
  BB0 {
    _0 <- field_00 _1;
    return _0
  }
  
end
module C02Nested_NestedClosure_Closure0_Type
  use prelude.Int8
  use prelude.Borrow
  type c02nested_nestedclosure_closure0  =
    | C02Nested_NestedClosure_Closure0 bool
    
end
module C02Nested_NestedClosure_Closure0
  use prelude.Int8
  use C02Nested_NestedClosure_Closure0_Type as C02Nested_NestedClosure_Closure0
  use prelude.Borrow
  use C02Nested_NestedClosure_Closure0_Closure0_Type as C02Nested_NestedClosure_Closure0_Closure0
  predicate resolve0 [#"../02_nested.rs" 6 18 6 20] (_1 : C02Nested_NestedClosure_Closure0_Closure0.c02nested_nestedclosure_closure0_closure0)
    
   =
    [#"../02_nested.rs" 1 0 1 0] true
  val closure00 [#"../02_nested.rs" 6 18 6 20] (_1 : C02Nested_NestedClosure_Closure0_Closure0.c02nested_nestedclosure_closure0_closure0) : bool
    
  let function field_00 [#"../02_nested.rs" 5 14 5 16] (self : C02Nested_NestedClosure_Closure0.c02nested_nestedclosure_closure0) : bool
    
   = [@vc:do_not_keep_trace] [@vc:sp]
    [#"../02_nested.rs" 1 0 1 0] let C02Nested_NestedClosure_Closure0.C02Nested_NestedClosure_Closure0 a = self in a
  let rec cfg c02Nested_NestedClosure_Closure0 [#"../02_nested.rs" 5 14 5 16] [@cfg:stackify] [@cfg:subregion_analysis] (_1 : C02Nested_NestedClosure_Closure0.c02nested_nestedclosure_closure0) : bool
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : bool;
  var _1 : C02Nested_NestedClosure_Closure0.c02nested_nestedclosure_closure0 = _1;
  var omg : C02Nested_NestedClosure_Closure0_Closure0.c02nested_nestedclosure_closure0_closure0;
  {
    goto BB0
  }
  BB0 {
    omg <- ([#"../02_nested.rs" 6 18 6 22] C02Nested_NestedClosure_Closure0_Closure0.C02Nested_NestedClosure_Closure0_Closure0 ([#"../02_nested.rs" 6 18 6 22] field_00 _1));
    _0 <- ([#"../02_nested.rs" 7 8 7 15] let () = [#"../02_nested.rs" 7 8 7 15] () in closure00 ([#"../02_nested.rs" 7 8 7 13] omg));
    goto BB1
  }
  BB1 {
    assume { resolve0 omg };
    return _0
  }
  
end
module C02Nested_NestedClosure
  use prelude.Int8
  use C02Nested_NestedClosure_Closure0_Type as C02Nested_NestedClosure_Closure0
  predicate resolve0 [#"../02_nested.rs" 5 14 5 16] (_1 : C02Nested_NestedClosure_Closure0.c02nested_nestedclosure_closure0)
    
   =
    [#"../02_nested.rs" 1 0 1 0] true
  use prelude.Borrow
  val closure00 [#"../02_nested.rs" 5 14 5 16] (_1 : C02Nested_NestedClosure_Closure0.c02nested_nestedclosure_closure0) : bool
    
  let rec cfg nested_closure [#"../02_nested.rs" 3 0 3 23] [@cfg:stackify] [@cfg:subregion_analysis] (_1 : ()) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a : bool;
  var _a : bool;
  var _4 : C02Nested_NestedClosure_Closure0.c02nested_nestedclosure_closure0;
  {
    goto BB0
  }
  BB0 {
    a <- ([#"../02_nested.rs" 4 12 4 16] [#"../02_nested.rs" 4 12 4 16] true);
    _4 <- ([#"../02_nested.rs" 5 13 8 6] C02Nested_NestedClosure_Closure0.C02Nested_NestedClosure_Closure0 ([#"../02_nested.rs" 5 13 8 6] a));
    _a <- ([#"../02_nested.rs" 5 13 8 8] let () = [#"../02_nested.rs" 5 13 8 8] () in closure00 ([#"../02_nested.rs" 5 13 8 6] _4));
    goto BB1
  }
  BB1 {
    assume { resolve0 _4 };
    _0 <- ([#"../02_nested.rs" 3 24 9 1] ());
    return _0
  }
  
end
