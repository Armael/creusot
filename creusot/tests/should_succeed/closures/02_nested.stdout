module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module C02Nested_NestedClosure_Closure0_Closure0_Interface
  use prelude.Prelude
  type c02nested_nestedclosure_closure0_closure0  = 
    | C02Nested_NestedClosure_Closure0_Closure0 bool
    
  function c02nested_nestedclosure_closure0_closure0_0 (self : c02nested_nestedclosure_closure0_closure0) : bool
  val c02nested_nestedclosure_closure0_closure0_0 (self : c02nested_nestedclosure_closure0_closure0) : bool
    ensures { result = c02nested_nestedclosure_closure0_closure0_0 self }
    
  axiom c02nested_nestedclosure_closure0_closure0_0_acc : forall a : bool . c02nested_nestedclosure_closure0_closure0_0 (C02Nested_NestedClosure_Closure0_Closure0 a : c02nested_nestedclosure_closure0_closure0) = a
  predicate precondition (_1' : c02nested_nestedclosure_closure0_closure0) (_2' : ()) = 
    true
  predicate postcondition (_1' : c02nested_nestedclosure_closure0_closure0) (_2' : ()) (result : bool) = 
    true
  predicate postcondition_mut (_1' : borrowed c02nested_nestedclosure_closure0_closure0) (_2' : ()) (result : bool) = 
    true && true
  predicate postcondition_once (_1' : c02nested_nestedclosure_closure0_closure0) (_2' : ()) (result : bool) = 
    true
  predicate resolve (_1' : c02nested_nestedclosure_closure0_closure0) = 
    (fun x -> true) (c02nested_nestedclosure_closure0_closure0_0 _1') && true
  val c02Nested_NestedClosure_Closure0_Closure0 (_1' : c02nested_nestedclosure_closure0_closure0) (_2' : ()) : bool
end
module C02Nested_NestedClosure_Closure0_Closure0
  use prelude.Prelude
  type c02nested_nestedclosure_closure0_closure0  = 
    | C02Nested_NestedClosure_Closure0_Closure0 bool
    
  function c02nested_nestedclosure_closure0_closure0_0 (self : c02nested_nestedclosure_closure0_closure0) : bool
  val c02nested_nestedclosure_closure0_closure0_0 (self : c02nested_nestedclosure_closure0_closure0) : bool
    ensures { result = c02nested_nestedclosure_closure0_closure0_0 self }
    
  axiom c02nested_nestedclosure_closure0_closure0_0_acc : forall a : bool . c02nested_nestedclosure_closure0_closure0_0 (C02Nested_NestedClosure_Closure0_Closure0 a : c02nested_nestedclosure_closure0_closure0) = a
  let rec cfg c02Nested_NestedClosure_Closure0_Closure0 (_1' : c02nested_nestedclosure_closure0_closure0) (_2' : ()) : bool
    
   = 
  var _0 : bool;
  var _1 : c02nested_nestedclosure_closure0_closure0;
  {
    _1 <- _1';
    goto BB0
  }
  BB0 {
    assume { (fun x -> true) _0 };
    _0 <- c02nested_nestedclosure_closure0_closure0_0 _1;
    assume { (fun x -> true) _1 };
    return _0
  }
  
end
module Core_Ops_Function_FnOnce_Output
  type self   
  type args   
  type output   
end
module Core_Ops_Function_Fn_Call_Interface
  type self   
  type args   
  use prelude.Prelude
  clone Core_Ops_Function_FnOnce_Output as Output0 with type self = self, type args = args
  val call [@cfg:stackify] (self : self) (args : args) : Output0.output
    requires {false}
    
end
module Core_Ops_Function_Fn_Call
  type self   
  type args   
  use prelude.Prelude
  clone Core_Ops_Function_FnOnce_Output as Output0 with type self = self, type args = args
  val call [@cfg:stackify] (self : self) (args : args) : Output0.output
    requires {false}
    
end
module C02Nested_NestedClosure_Closure0_Interface
  use prelude.Prelude
  type c02nested_nestedclosure_closure0  = 
    | C02Nested_NestedClosure_Closure0 bool
    
  function c02nested_nestedclosure_closure0_0 (self : c02nested_nestedclosure_closure0) : bool
  val c02nested_nestedclosure_closure0_0 (self : c02nested_nestedclosure_closure0) : bool
    ensures { result = c02nested_nestedclosure_closure0_0 self }
    
  axiom c02nested_nestedclosure_closure0_0_acc : forall a : bool . c02nested_nestedclosure_closure0_0 (C02Nested_NestedClosure_Closure0 a : c02nested_nestedclosure_closure0) = a
  predicate precondition (_1' : c02nested_nestedclosure_closure0) (_2' : ()) = 
    true
  predicate postcondition (_1' : c02nested_nestedclosure_closure0) (_2' : ()) (result : bool) = 
    true
  predicate postcondition_mut (_1' : borrowed c02nested_nestedclosure_closure0) (_2' : ()) (result : bool) = 
    true && true
  predicate postcondition_once (_1' : c02nested_nestedclosure_closure0) (_2' : ()) (result : bool) = 
    true
  predicate resolve (_1' : c02nested_nestedclosure_closure0) = 
    (fun x -> true) (c02nested_nestedclosure_closure0_0 _1') && true
  val c02Nested_NestedClosure_Closure0 (_1' : c02nested_nestedclosure_closure0) (_2' : ()) : bool
end
module C02Nested_NestedClosure_Closure0
  use prelude.Prelude
  type c02nested_nestedclosure_closure0  = 
    | C02Nested_NestedClosure_Closure0 bool
    
  function c02nested_nestedclosure_closure0_0 (self : c02nested_nestedclosure_closure0) : bool
  val c02nested_nestedclosure_closure0_0 (self : c02nested_nestedclosure_closure0) : bool
    ensures { result = c02nested_nestedclosure_closure0_0 self }
    
  axiom c02nested_nestedclosure_closure0_0_acc : forall a : bool . c02nested_nestedclosure_closure0_0 (C02Nested_NestedClosure_Closure0 a : c02nested_nestedclosure_closure0) = a
  clone C02Nested_NestedClosure_Closure0_Closure0_Interface as Closure00 with axiom .
  let rec cfg c02Nested_NestedClosure_Closure0 (_1' : c02nested_nestedclosure_closure0) (_2' : ()) : bool = 
  var _0 : bool;
  var _1 : c02nested_nestedclosure_closure0;
  var omg_2 : Closure00.c02nested_nestedclosure_closure0_closure0;
  var _3 : bool;
  var _4 : Closure00.c02nested_nestedclosure_closure0_closure0;
  var _5 : ();
  {
    _1 <- _1';
    goto BB0
  }
  BB0 {
    _3 <- c02nested_nestedclosure_closure0_0 _1;
    assume { (fun x -> true) _1 };
    omg_2 <- Closure00.C02Nested_NestedClosure_Closure0_Closure0 _3;
    _4 <- omg_2;
    assume { (fun x -> true) omg_2 };
    _5 <- ();
    _0 <- Closure00.c02Nested_NestedClosure_Closure0_Closure0 _4 _5;
    goto BB1
  }
  BB1 {
    return _0
  }
  
end
module C02Nested_NestedClosure_Interface
  val nested_closure [@cfg:stackify] () : ()
end
module C02Nested_NestedClosure
  use prelude.Prelude
  clone C02Nested_NestedClosure_Closure0_Interface as Closure00 with axiom .
  let rec cfg nested_closure [@cfg:stackify] () : () = 
  var _0 : ();
  var a_1 : bool;
  var a_2 : bool;
  var _3 : Closure00.c02nested_nestedclosure_closure0;
  var _4 : Closure00.c02nested_nestedclosure_closure0;
  var _5 : bool;
  var _6 : ();
  {
    goto BB0
  }
  BB0 {
    a_1 <- true;
    _5 <- a_1;
    assume { (fun x -> true) a_1 };
    _4 <- Closure00.C02Nested_NestedClosure_Closure0 _5;
    _3 <- _4;
    assume { (fun x -> true) _4 };
    _6 <- ();
    a_2 <- Closure00.c02Nested_NestedClosure_Closure0 _3 _6;
    goto BB1
  }
  BB1 {
    assume { (fun x -> true) a_2 };
    _0 <- ();
    return _0
  }
  
end
